src/warpychain/poller.go

```diff
@@ -0,0 +1,137 @@
+
+	self.Task = task.NewTask(config, "poller").
+		WithRepeatedSubtaskFunc(config.WarpyChainer.PollerInterval, self.poll).
+		WithOnAfterStop(func() {
+			close(self.Output)
+		})
+	return
+}
+
+func (self *Poller) WithMonitor(monitor monitoring.Monitor) *Poller {
+	self.monitor = monitor
+	return self
+}
+
+func (self *Poller) WithDB(db *gorm.DB) *Poller {
+	self.db = db
+	return self
+}
+
+func (self *Poller) WithBackoff(maxElapsedTime, maxInterval time.Duration) *Poller {
+	self.maxElapsedTime = maxElapsedTime
+	self.maxInterval = maxInterval
+	return self
+}
+
+func (self *Poller) WithWarpyChainClient(client *WarpyChainClient) *Poller {
+	self.warpyChainClient = client
+	return self
+}
+
+func (self *Poller) poll() (repeat bool, err error) {
+	self.Log.Info("Polling new block")
+	ctx, cancel := context.WithTimeout(self.Ctx, self.Config.WarpyChainer.PollerTimeout)
+	defer cancel()
+
+	txs, lastSyncedHeight, err := self.processBlockAndGetRedStoneTx(ctx)
+
+	if err != nil {
+		self.Log.WithField("height", lastSyncedHeight).WithError(err).Error("Could not process block")
+		self.monitor.GetReport().WarpyChainer.Errors.PollerProcessingBlockError.Inc()
+		return
+	}
+
+	if len(txs) == 0 {
+		return
+	}
+
+	for _, tx := range txs {
+		self.Output <- tx
+	}
+
+	self.Log.WithField("height", lastSyncedHeight).Info("Block processed")
+	self.monitor.GetReport().WarpyChainer.State.PollerBlocksSynced.Inc()
+	repeat = true
+	return
+}
+
+func (self *Poller) processBlockAndGetRedStoneTx(ctx context.Context) (txs []*types.Transaction, lastSyncedHeight int64, err error) {
+
+	err = task.NewRetry().
+		WithContext(self.Ctx).
+		WithMaxElapsedTime(self.maxElapsedTime).
+		WithMaxInterval(self.maxInterval).
+		WithAcceptableDuration(self.maxInterval * 2).
+		WithOnError(func(err error, isDurationAcceptable bool) error {
+			if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+				// Stopping permanently
+				return backoff.Permanent(err)
+			}
+
+			self.Log.WithError(err).Warn("Could not process block, retrying...")
+			return err
+		}).
+		Run(func() error {
+			txs = []*types.Transaction{}
+			lastSyncedHeight, err := self.warpyChainClient.getLastSyncedBlockHeight(ctx)
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get last synced height")
+				return err
+			}
+
+			self.Log.WithField("height", lastSyncedHeight).Debug("Checking for new block height")
+			blockInfo, err := self.warpyChainClient.getBlockInfo(lastSyncedHeight + 1)
+			if err != nil {
+				self.Log.WithError(err).Warn("Could not get block info")
+				return err
+			}
+			lastSyncedHeight = lastSyncedHeight + 1
+			transactions := blockInfo.Transactions()
+			for i := 0; i < len(transactions); i++ {
+				isRedStoneTx := self.warpyChainClient.checkForRedStoneTx(transactions[i], ctx)
+				if isRedStoneTx {
+					self.monitor.GetReport().WarpyChainer.State.PollerRedStoneTransactionsNum.Inc()
+					txs = append(txs, transactions[i])
```

Comment:
If it was me (and I'm lazy and dumb), I would simply call here https://github.com/warp-contracts/syncer/pull/79/files#diff-3e7ad67e5c7273756828ae1c902de69240341b3d887e6c0092aa57b7dcff6ea0R73 :-) - instead of writing another pile of code in a separate `writer.go`